File: backend/src/main/java/br/com/calciolari/datahub/imports/infrastructure/interpdv/qrp/QrpContainerReader.java
package br.com.calciolari.datahub.imports.infrastructure.interpdv.qrp;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * Locates embedded EMF pages inside a QRP blob.
 *
 * <p>Evidence from PoC {@code isEmfAt} / {@code findEmfPages}:
 * <ul>
 *   <li>EMR_HEADER type {@code 1} at offset {@code i}</li>
 *   <li>ASCII {@code " EMF"} (0x20 0x45 0x4D 0x46) at {@code i+40}</li>
 *   <li>page byte length {@code nBytes} LE u32 at {@code i+48}</li>
 *   <li>require {@code nBytes > 80} and {@code i + nBytes <= file length}</li>
 * </ul>
 */
public final class QrpContainerReader {

	static final int EMR_HEADER = 1;
	static final int EMF_SIGNATURE_OFFSET = 40;
	static final int EMF_NBYTES_OFFSET = 48;
	static final int MIN_HEADER_WINDOW = 56;
	static final int MIN_PAGE_BYTES = 80;

	/** ASCII {@code " EMF"} */
	private static final byte[] EMF_SIGNATURE = {0x20, 0x45, 0x4D, 0x46};

	private final int maxPages;

	public QrpContainerReader() {
		this(512);
	}

	public QrpContainerReader(int maxPages) {
		if (maxPages < 1) {
			throw new IllegalArgumentException("maxPages must be >= 1");
		}
		this.maxPages = maxPages;
	}

	public List<EmfPage> findEmfPages(byte[] bytes) {
		Objects.requireNonNull(bytes, "bytes");
		ByteBuffer view = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN);
		List<EmfPage> pages = new ArrayList<>();
		for (int i = 0; i + MIN_HEADER_WINDOW < bytes.length; i++) {
			if (!isEmfAt(bytes, view, i)) {
				continue;
			}
			long nBytesLong = Integer.toUnsignedLong(view.getInt(i + EMF_NBYTES_OFFSET));
			if (nBytesLong > Integer.MAX_VALUE) {
				continue;
			}
			int nBytes = (int) nBytesLong;
			if (nBytes > MIN_PAGE_BYTES && i <= bytes.length - nBytes) {
				pages.add(new EmfPage(i, nBytes));
				if (pages.size() >= maxPages) {
					break;
				}
				i += nBytes - 1;
			}
		}
		return List.copyOf(pages);
	}

	static boolean isEmfAt(byte[] bytes, ByteBuffer view, int i) {
		if (i < 0 || i + MIN_HEADER_WINDOW > bytes.length) {
			return false;
		}
		if (view.getInt(i) != EMR_HEADER) {
			return false;
		}
		return bytes[i + EMF_SIGNATURE_OFFSET] == EMF_SIGNATURE[0]
				&& bytes[i + EMF_SIGNATURE_OFFSET + 1] == EMF_SIGNATURE[1]
				&& bytes[i + EMF_SIGNATURE_OFFSET + 2] == EMF_SIGNATURE[2]
				&& bytes[i + EMF_SIGNATURE_OFFSET + 3] == EMF_SIGNATURE[3];
	}
}

File: backend/src/Calciolari.DataHub/Imports/Infrastructure/InterPdv/Qrp/QrpContainerReader.cs
using System.Buffers.Binary;

namespace Calciolari.DataHub.Imports.Infrastructure.InterPdv.Qrp;

/// <summary>
/// Locates embedded EMF pages inside a QRP blob.
/// Evidence from PoC isEmfAt / findEmfPages.
/// </summary>
public sealed class QrpContainerReader
{
    internal const int EmrHeader = 1;
    internal const int EmfSignatureOffset = 40;
    internal const int EmfNbytesOffset = 48;
    internal const int MinHeaderWindow = 56;
    internal const int MinPageBytes = 80;

    private static readonly byte[] EmfSignature = [0x20, 0x45, 0x4D, 0x46];

    private readonly int _maxPages;

    public QrpContainerReader()
        : this(512)
    {
    }

    public QrpContainerReader(int maxPages)
    {
        if (maxPages < 1)
        {
            throw new ArgumentOutOfRangeException(nameof(maxPages), "maxPages must be >= 1");
        }

        _maxPages = maxPages;
    }

    public IReadOnlyList<EmfPage> FindEmfPages(byte[] bytes)
    {
        ArgumentNullException.ThrowIfNull(bytes);
        var pages = new List<EmfPage>();
        for (var i = 0; i + MinHeaderWindow < bytes.Length; i++)
        {
            if (!IsEmfAt(bytes, i))
            {
                continue;
            }

            var nBytesLong = BinaryPrimitives.ReadUInt32LittleEndian(bytes.AsSpan(i + EmfNbytesOffset, 4));
            if (nBytesLong > int.MaxValue)
            {
                continue;
            }

            var nBytes = (int)nBytesLong;
            if (nBytes > MinPageBytes && i <= bytes.Length - nBytes)
            {
                pages.Add(EmfPage.Create(i, nBytes));
                if (pages.Count >= _maxPages)
                {
                    break;
                }

                i += nBytes - 1;
            }
        }

        return pages;
    }

    internal static bool IsEmfAt(byte[] bytes, int i)
    {
        if (i < 0 || i + MinHeaderWindow > bytes.Length)
        {
            return false;
        }

        if (BinaryPrimitives.ReadInt32LittleEndian(bytes.AsSpan(i, 4)) != EmrHeader)
        {
            return false;
        }

        return bytes[i + EmfSignatureOffset] == EmfSignature[0]
               && bytes[i + EmfSignatureOffset + 1] == EmfSignature[1]
               && bytes[i + EmfSignatureOffset + 2] == EmfSignature[2]
               && bytes[i + EmfSignatureOffset + 3] == EmfSignature[3];
    }
}
